akka-stream-tests/src/test/scala/akka/stream/impl/fusing/InterpreterSupervisionSpec.scala

```diff
@@ -0,0 +1,656 @@
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(2)
+      lastEvents() should be(Set(OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(10) // boom
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(4)
+      lastEvents() should be(Set(OnNext(3))) // 1 + 2
+    }
+
+    "restart when Scan throws" in new TestSetup(Seq(
+      Scan(1, (acc: Int, x: Int) ⇒ if (x == 10) throw TE else acc + x, restartingDecider))) {
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(2)
+      lastEvents() should be(Set(OnNext(1)))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(10) // boom
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(4)
+      lastEvents() should be(Set(OnNext(1))) // starts over again
+
+      downstream.requestOne()
+      lastEvents() should be(Set(RequestOne))
+      upstream.onNext(20)
+      lastEvents() should be(Set(OnNext(5))) // 1+4
+    }
+
+    "restart when Conflate `seed` throws" in new TestSetup(Seq(Conflate(
+      (in: Int) ⇒ if (in == 1) throw TE else in,
+      (agg: Int, x: Int) ⇒ agg + x,
+      restartingDecider))) {
+
+      lastEvents() should be(Set(RequestOne))
+
+      downstream.requestOne()
+      lastEvents() should be(Set.empty)
+
+      upstream.onNext(0)
+      lastEvents() should be(Set(OnNext(0), RequestOne))
+
+      upstream.onNext(1) // boom
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(2)
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(10)
+      lastEvents() should be(Set(RequestOne))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(12))) // note that 1 has been discarded
+
+      downstream.requestOne()
+      lastEvents() should be(Set.empty)
+    }
+
+    "restart when Conflate `aggregate` throws" in new TestSetup(Seq(Conflate(
+      (in: Int) ⇒ in,
+      (agg: Int, x: Int) ⇒ if (x == 2) throw TE else agg + x,
+      restartingDecider))) {
+
+      lastEvents() should be(Set(RequestOne))
+
+      downstream.requestOne()
+      lastEvents() should be(Set.empty)
+
+      upstream.onNext(0)
+      lastEvents() should be(Set(OnNext(0), RequestOne))
+
+      upstream.onNext(1)
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(2) // boom
+      lastEvents() should be(Set(RequestOne))
+
+      upstream.onNext(10)
+      lastEvents() should be(Set(RequestOne))
+
+      downstream.requestOne()
+      lastEvents() should be(Set(OnNext(10))) // note that 1 and 2 has been discarded
+
+      downstream.requestOne()
+      lastEvents() should be(Set.empty)
+
+      upstream.onNext(4)
+      lastEvents() should be(Set(OnNext(4), RequestOne))
+
+      downstream.cancel()
+      lastEvents() should be(Set(Cancel))
+    }
+
```

Comment:
@drewhk I have added a lot of tests in this InterpreterSupervisionSpec.scala file. Above this line are tests that I think work as expected. I have some troubles below this line, and would appreciate input. See FIXMEs below...